contracts/src/v0.1/RequestResponseCoordinator.sol

```diff
@@ -521,4 +481,190 @@ contract RequestResponseCoordinator is
+    ) internal returns (bool) {
+        if (!s_oracles[msg.sender]) {
+            revert UnregisteredOracleFulfillment(msg.sender);
+        }
+
+        bytes32 commitment = s_requestCommitments[requestId];
+        if (commitment == 0) {
+            revert NoCorrespondingRequest();
+        }
+
+        if (
+            commitment !=
+            keccak256(abi.encode(requestId, rc.blockNum, rc.accId, rc.callbackGasLimit, rc.sender))
+        ) {
+            revert IncorrectCommitment();
+        }
+
+        delete s_requestCommitments[requestId];
+
+        return true;
+    }
+
+    function fulfill(bytes memory resp, RequestCommitment memory rc) internal returns (bool) {
+        // Call with explicitly the amount of callback gas requested
+        // Important to not let them exhaust the gas budget and avoid oracle payment.
+        // Do not allow any non-view/non-pure coordinator functions to be called
+        // during the consumers callback code via reentrancyLock.
+        // Note that callWithExactGas will revert if we do not have sufficient gas
+        // to give the callee their requested amount.
+        s_config.reentrancyLock = true;
+        bool success = callWithExactGas(rc.callbackGasLimit, rc.sender, resp);
+        s_config.reentrancyLock = false;
+        return success;
+    }
+
+    function pay(
+        RequestCommitment memory rc,
+        bool isDirectPayment,
+        uint256 startGas
+    ) internal returns (uint256 payment) {
+        // We want to charge users exactly for how much gas they use in their callback.
+        // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+        // decrement the account balance and increment the oracles withdrawable balance.
+        // We also add the flat KLAY fee to the payment amount.
+        // Its specified in millionths of KLAY, if s_config.fulfillmentFlatFeeKlayPPM = 1
+        // 1 KLAY / 1e6 = 1e18 pebs / 1e6 = 1e12 pebs.
+        (uint256 balance, uint64 reqCount, , ) = s_prepayment.getAccount(rc.accId);
+
+        if (isDirectPayment) {
+            payment = balance;
+        } else {
+            payment = calculatePaymentAmount(
+                startGas,
+                s_config.gasAfterPaymentCalculation,
+                getFeeTier(reqCount)
+            );
+        }
+
+        s_prepayment.chargeFee(rc.accId, payment, msg.sender);
+
+        return payment;
+    }
+
+    function fulfillDataRequestUint256(
+        uint256 requestId,
+        uint256 response,
+        RequestCommitment memory rc,
+        bool isDirectPayment
+    ) external nonReentrant returns (uint256) {
+        uint256 startGas = gasleft();
+        validateDataResponse(rc, requestId);
+        RequestResponseConsumerBase rr;
+        bytes memory resp = abi.encodeWithSelector(
+            rr.rawFulfillDataRequestUint256.selector,
+            requestId,
+            response
+        );
+        bool success = fulfill(resp, rc);
```

Comment:
Theoretically, we could merge the `fulfill` & `pay` functions together, but let's keep them separated for now. Later, we should measure the gas between those two approaches and ensure security.